src/key.h

```diff
@@ -128,6 +128,9 @@ class CKey
     //! Derive BIP32 child key.
     bool Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const ChainCode& cc) const;
 
+    // computes an EC Diffie-Hellman 256bit secret
+    bool ComputeECDHSecret(const CPubKey& pubkey, CPrivKey& secret_out) const;
```

Comment:
Extract `ComputeECDHSecret` and make it a freestanding function `bool ComputeECDHSecret(const CKey& key, const CPubKey& pubkey, CPrivKey& secret_out)`?

Although it is convenient to add new BIP responsibilities directly to `CKey`, I would recommend against adding more responsibilities to `CKey` than it already has.

Reply:
I think this makes sense to be a member function (in align with `Derive` and other member functions of `CKey`).

Reply:
"*A function that can be simply, elegantly, and efficiently implemented as a freestanding function (that is, as a nonmember function) should be implemented outside the class.*" (Stroustrup) is a very commonly used and well documented best practice in C++ development.

See [C.4. Make a function a member only if it needs direct access to the representation of a class](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c4-make-a-function-a-member-only-if-it-needs-direct-access-to-the-representation-of-a-class) for more info and guidelines.

```c++
bool ComputeECDHSecret(const CKey& key, const CPubKey& pubkey, CPrivKey& secret_out)
{
    secp256k1_pubkey pubkey_internal;
    if (!secp256k1_ec_pubkey_parse(secp256k1_context_sign, &pubkey_internal, pubkey.data(), pubkey.size())) {
        return false;
    }

    secret_out.resize(32);
    return secp256k1_ecdh(secp256k1_context_sign, &secret_out[0], &pubkey_internal, key.begin()) == 1;
}
```